### ISSUE TITLE
Strange error when typechecking.

### ISSUE THREAD

**Issue by zsszatmari** [facebook/flow#1021]
Hi!

I am trying to use flow together with the seamless-immutable package, and add a declaration so it would typecheck. I get strange error:

``` javascript
declare class ImmutableExt<T>
{
    merge(x : any, options : ?any) : Immutable<T>;
}
type Immutable<T> = T & ImmutableExt<T>;
declare function immutable<T>(x : T) : Immutable<T>;

type ProgressState = Immutable<{failedActions: Array<any>, succeededActions: Array<any>}>;

function progress(state /*: ?ProgressState */) /*: ProgressState */
{
    if (!state) {
        return immutable({failedActions: [], succeededActions: []});
    } else {
        return state;
    }

    // curiously, if I only have return state, without any if's, it typechecks
}
```

```
 33:    merge(x : any, options : ?any) : Immutable<T>;
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ property `merge`. Property not found in
 45:        return state;
              ^^^^^ object type


Found 1 error
```

**Comment by samwgoldman** [comment]
Maybe related to #582.

**Comment by samwgoldman** [comment]
This is still failing after the fix for #582.

**Comment by TrySound** [comment]
Looks like it fixed
https://flow.org/try/#0CYUwxgNghgTiAEkoGdnwJIFtMFcAuUARhCAKIAeeAPACoB8AUAN4PxvyYgwDmIAFOXgAueFAB2ATwA08APYAHPAEtZYtCID84iQEphGbPiIladANwMAvgzwT5CLLgLEQp+AF54NeADIDT4zJKUwtQJDh4ADMcMTBlVXglQ2cTegF9Gj0RRyMXEIYbOwQABRhZbjhUAGUCPARPHJTXJkioJRJgAEE4lTURTpgYKAkqbToZZBwwMBAQUC6e1WR+weHRyTpLcwLo2PixeHkyipBUPmRahAB6ACpNUvLK5BqoOvgbq71bkQeT6sv3ldmKx2EpIvA+ABCC6vEB6FjsRHwOB4HAwA5JAIuPgtNodbr7ZbwADaAF0JlMZnM5gTekSyZYdBZEZZ4CAIMgEAikWwUWiDjC6sz2NYQWwrldEGiVDhkBBpIlweg5GJ5fAABZQABuCD56PggpAMgA7ko8OrZPhRJJFQByZAyM3wWz2MDq8AAa2QViAA
